Use strongly-typed types.NodeName for a node name

justinsb opened the pull request:
We had another bug where we confused the hostname with the NodeName.

Also, if we want to use different values for the Node.Name (which is
an important step for making installation easier), we need to keep
better control over this.

A tedious but mechanical commit therefore, to change all uses of the
node name to use types.NodeName

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29048)

<!-- Reviewable:end -->

justinsb committed:
Use strongly-typed types.NodeName for a node name

We had another bug where we confused the hostname with the NodeName.

To avoid this happening again, and to make the code more
self-documenting, we use types.NodeName (a typedef alias for string)
whenever we are referring to the Node.Name.

A tedious but mechanical commit therefore, to change all uses of the
node name to use types.NodeName

Also clean up some of the (many) places where the NodeName is referred
to as a hostname (not true on AWS), or an instanceID (not true on GCE),
etc.

justinsb commented:
WIP mainly for discussion of where NodeName should live... maybe `pkg/api`?

This is frustratingly big, but it has two benefits:

* Fix bugs like #28643 and prevent them happening again
* Make the code more self-documenting - it is really unobvious what "instance string" or "host string" is, so it is unsurprising that we make mistakes and it is hard for people writing new cloudproviders (for example).

lavalamp commented:
Plan sounds good, we should do more things like this. pkg/types seems like a good home. Need to make sure the swagger spec doesn't change (still lists it as a 'string').

@caesarxuchao can you make a pass over this?

Let's get this mostly reviewed, then warn us before you rebase so we can get to it quickly after that so you don't have to rebase again.

caesarxuchao commented on pkg/api/types.go:
```diff
@@ -2427,7 +2427,7 @@ type SerializedReference struct {
 	// Component from which the event is generated.
 	Component string `json:"component,omitempty"`
-	// Host name on which the event is generated.
+	// Node name on which the event is generated.
```
Is this change intentional?

justinsb replied:
Yes; I may be wrong, but I believe this is a Node.Name, not a host name.  If we are cross-referencing an event to a node, we should look up a Node.Name, not a host name.  We don't even publish the host name on a Node (except when it == the node name), so it would be wrong to publish with the hostname.

I don't think we can fix the json (that's schema), but we can at least fix the comment.

caesarxuchao replied:
I did a search in the code base and it seems we never set the EventSource.Host. If that's true, we might be able to change the schema.

> We don't even publish the host name on a Node

Do you mean we can't find the host name from the Node api?

caesarxuchao replied:
Oh I only searched inside pkg/ and Host is set in cmd/. I agree it should actually be termed as node name.

caesarxuchao replied:
Could you also change the comment in pkg/api/v1/types.go?

justinsb replied:
Changed :-)

caesarxuchao commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -442,7 +442,7 @@ func (c *Cloud) AddSSHKeyToAllInstances(user string, keyData []byte) error {
 
 // CurrentNodeName returns the name of the current node
-func (c *Cloud) CurrentNodeName(hostname string) (string, error) {
+func (c *Cloud) CurrentNodeName(hostname string) (types.NodeName, error) {
```
Can we remove the input hostname?

caesarxuchao replied:
Nvm, obviously it's implementing an interface..

pmorie replied:
pedantic nit: i would be cool with more interface checks for the interfaces `Cloud` implements

justinsb replied:
@pmorie by that do you mean `var _ Interface = &Impl{}` or something else.  I'm inclined to agree with you, though not necessarily in this PR!

I think the original reasoning behind not just using checked casting was that in theory another object might implement one of the cloud provider interfaces.  I don't think anyone actually does that though.

caesarxuchao commented:
@justinsb sorry for the delay. I found myself not understanding the difference between HostName and NodeName. After skimming some of the changed files, it seems to me InstanceName is the same as NodeName, right? Could you explain what is HostName and when we should use HostName instead of NodeName? Thanks.

pmorie commented:
I support this patch.

I think it would be good to articulate the information that @caesarxuchao is asking for somewhere int the code.  It's a good question, and deserves explanation in a comment.

justinsb commented:
I can try!  I think we've definitely fallen into a lot of confusion in the past here, and I'm by no means immune.  Most of these changes were driven by the golang type checker, so that's a good sign for the future.

* Node.Name is the Name field of the Node in the API.
* Hostname is the hostname of the local machine (from `uname -n`).  However, some components allow you to pass in a `--hostname-override` flag which will override this in most places.  In the absence of anything more meaningful, kubelet will use Hostname as the Node.Name when it creates the Node.
* InstanceName is I believe only in the GCE provider.  It is the Name of an Instance object in the GCE API.  On GCE, Instance.Name becomes the Hostname, and thus it makes sense also to use it as the Node.Name.  But that's GCE specific, and it is up to the cloudprovider how to do this mapping.  I wrapped these conversions in helper functions `mapInstanceToNodeName` so that it would be done in one place rather than having string casts everywhere.

I could probably take this further, but this PR is already getting way too big.  For example, the EnsureLoadBalancer function is current taking an array of `hosts []string`.  This is (unsurprisingly) actually `nodeNames []types.NodeName`.  (But I think we might actually want to make that `nodes []*api.Node` anyway.)

It is a pity that I can't make `Node.Name` be of type NodeName (as every cast is suspicious), but it is actually in ObjectMeta.  Maybe we could adjust the wrapper layers, but that's also for another PR I think...

justinsb commented:
Edit: and forgot to ask ... any thoughts on where to put the comment?  How about above the NodeName type itself?

saad-ali commented:
> I support this patch.

+1

caesarxuchao commented:
For future reference, this is code that populates the Node name: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L631, Node.Name is populated from kcfg.NodeName.

caesarxuchao commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -827,15 +827,15 @@ func (c *Cloud) NodeAddresses(name string) ([]api.NodeAddress, error) {
-// ExternalID returns the cloud provider ID of the specified instance (deprecated).
-func (c *Cloud) ExternalID(name string) (string, error) {
-	if c.selfAWSInstance.nodeName == name {
+// ExternalID returns the cloud provider ID of the node with the specified Name (deprecated).
```
s/specified Name/specified nodeName? or is it an intentional abbreviation?

caesarxuchao commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1313,8 +1313,8 @@ func (c *Cloud) buildSelfAWSInstance() (*awsInstance, error) {
 
-// Gets the awsInstance with node-name nodeName, or the 'self' instance if nodeName == ""
-func (c *Cloud) getAwsInstance(nodeName string) (*awsInstance, error) {
+// Gets the awsInstance with for the node with the specified Name Name, or the 'self' instance if nodeName == ""
```
s/Name Name/Name

caesarxuchao commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -188,8 +189,15 @@ func ipAddress(name string) (net.IP, error) {
+	return string(nodeName)
+}
+
 // ExternalID returns the cloud provider ID of the specified instance (deprecated).
```
Could you update the comment as you did for the aws and gce?

caesarxuchao commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -212,12 +220,12 @@ func (c *MesosCloud) ExternalID(instance string) (string, error) {
 }
 
 // InstanceID returns the cloud provider ID of the specified instance.
```
Update the comment?

caesarxuchao commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -212,12 +220,12 @@ func (c *MesosCloud) ExternalID(instance string) (string, error) {
 	return "", nil
 }
 
 // InstanceType returns the type of the specified instance.
```
Update the comment?

caesarxuchao commented on pkg/cloudprovider/providers/mesos/mesos.go:
```diff
@@ -274,7 +282,8 @@ func (c *MesosCloud) ListWithoutKubelet() ([]string, error) {
 }
 
 // NodeAddresses returns the addresses of the specified instance.
```
Update the comment?

caesarxuchao commented on pkg/cloudprovider/providers/ovirt/ovirt.go:
```diff
@@ -149,7 +150,8 @@ func (v *OVirtCloud) Routes() (cloudprovider.Routes, bool) {
 }
 
 // NodeAddresses returns the NodeAddresses of a particular machine instance
```
Update the comment?

caesarxuchao commented on pkg/kubelet/kubelet.go:
```diff
@@ -504,7 +504,7 @@ func NewMainKubelet(
 	klet.volumeManager, err = volumemanager.NewVolumeManager(
 		enableControllerAttachDetach,
-		hostname,
+		nodeName,
```
Is this where you fix the bug #28643?

caesarxuchao replied:
Nevermind.

caesarxuchao commented on pkg/volume/aws_ebs/attacher.go:
```diff
@@ -51,7 +52,7 @@ func (plugin *awsElasticBlockStorePlugin) NewAttacher() (volume.Attacher, error)
 }
 
-func (attacher *awsElasticBlockStoreAttacher) Attach(spec *volume.Spec, hostName string) (string, error) {
+func (attacher *awsElasticBlockStoreAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
```
Oh I got it, it's this Attach and AttachDisk that directly fix the #28643.

caesarxuchao commented:
Thanks, @justinsb. LGTM other than a few nits about the comments.

jingxu97 commented:
Fixes #29048

caesarxuchao commented:
@justinsb could you rebase and update the comments?

justinsb commented:
@caesarxuchao rebased, fixed comments, added big comment to guide future travelers here: https://github.com/justinsb/kubernetes/blob/4a37abf9b06fcd3b561a9072b6c9f1ab9162727f/pkg/types/nodename.go

caesarxuchao commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 17 of 55 files at r1, 51 of 51 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29048#-:-KOzXpFIN_aXDeosPSAt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
I'm learning how to use Reviewable, it seems there is a one discussion needs @pmorie to mark as satisfied. @pmorie could you help?

caesarxuchao commented:
@justinsb sorry you have to rebase again.

caesarxuchao commented:
@justinsb feel free to add the label yourself after rebase.

justinsb commented:
Rebased, re-self-reviewed it.  Will self-lgtm on passing e2e (unless I hear otherwise.)  Thanks for the reviews (which I know are painful!)

justinsb commented:
I'm actually thinking it might be a little obnoxious to merge this right now with everyone else also scrambling, given it will likely break a huge number of PRs.  I'll rebase again on Monday/Tuesday.

caesarxuchao commented:
Thanks @justinsb!

justinsb commented:
Tis the season (in the release cycle).  I've rebased, I believe tests should pass shortly.  Let's get this in now so we have plenty of time... and of course now I have more fixes I want to do based on this idea (mostly in the AWS provider, but also that LoadBalancer interface is still using `[]string`)!

caesarxuchao commented:
Reviewed 49 of 58 files at r3.
Review status: 70 of 78 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29048)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
LGTM again.Thanks @justinsb!

errordeveloper commented on cmd/kubeadm/app/node/csr.go:
```diff
@@ -20,6 +20,7 @@ import (
 	"fmt"
 	"os"
 
+	"github.com/golang/glog"
```
We don't use glog in kubeadm (or at least not yet), the default output format is too noisy.

errordeveloper commented on cmd/kubeadm/app/node/csr.go:
```diff
@@ -37,11 +39,16 @@ func PerformTLSBootstrap(s *kubeadmapi.KubeadmConfig, apiEndpoint string, caCert
 
+	// TODO: hostname == nodename doesn't hold on all clouds (AWS).
+	// But we don't have a cloudprovider, so we're stuck.
+	glog.Errorf("assuming that hostname is the same as NodeName")
```
I am not sure why you need to add this here, we need to open an issue to discuss this. My concern is that user has really no idea what the hell `hostname` vs `NodeName` means. I'll open a PR and an issue to fix this.
